### PR TITLE
`Process`: improve error messages inside `toBackend`

### PR DESCRIPTION
**Pull request by qaristote**
See #43.

### CODE REVIEW

File: smtlib-backends-process/src/SMTLIB/Backends/Process.hs [hunk 1]
```diff
@@ -161,7 +161,13 @@ toBackend :: Handle -> Backend
 toBackend handle =
   Backend $ \cmd -> do
     write handle cmd
-    toLazyByteString <$> continueNextLine (scanParen 0) mempty
+    toLazyByteString
+      <$> ( continueNextLine (scanParen 0) mempty
+              `X.catch` \ex ->
+                fail $
+                  "Error while reading solver's response: "
```
[the remark this answers is not in the document]
Comment (qaristote) [reply smtlib-backends-process/src/SMTLIB/Backends/Process.hs:168]: I'm all for it but I'm afraid the error message may become quite long, is that not a problem ?

**Comment by qaristote** [comment]
> Do you think a test should be added to specify the behaviour?

If you mean the behaviour of `scanParen`, yes this is probably a good idea.

> This would likely involve golden files and be quite heavy

I don't know what golden files are :see_no_evil: But yeah it'd be a bit tedious as we'd have to implement a dummy process to interact with the `toBackend` function. Or we could expose a `Process.read` function.

**Review by facundominguez — COMMENTED** [review]
`continueNextLine` could wrap the call the `hGetLine` with an exception handler that wraps the exception in another exception that contains the accumulated output (`toLazyByteString acc`).

**Comment by qaristote** [comment]
> `continueNextLine` could wrap the call the `hGetLine` with an exception handler that wraps the exception in another exception that contains the accumulated output (`toLazyByteString acc`).

So something like this ?
```haskell
continueNextLine f acc = do
      next <-
        BS.hGetLine (getStdout $ process handle)
          `X.catch` \ex ->
            error $
              "[smtlib-backends-process] Error while reading solver's response: "
                ++ show (ex :: X.IOException)
                ++ "\nThis happened after successfully reading:\n"
                ++ LBS.unpack (toLazyByteString acc)

      f (acc <> byteString next) next
```

**Comment by facundominguez** [comment]
Ah, that would work too. Here's what I was thinking.

```Haskell
continueNextLine f acc = do
      next <-
        BS.hGetLine (getStdout $ process handle)
          `X.catch` \ex ->
            throwIO (GetLineExceptionAfter ex (toLazyByteString acc))
      f (acc <> byteString next) next

data GetLineException = GetLineException IOException ByteString
  deriving Show
instance Exception (GetLineException) -- relies on default methods
```
then you would have to catch `GetLineException` when calling continueNextLine.
Or
```
continueNextLine f acc = do
      next <-
        BS.hGetLine (getStdout $ process handle)
          `X.catch` \ex ->
            throwIO (ex {ioe_description = ioe_description ex ++ ": " ++ show (toLazyByteString acc) })
      f (acc <> byteString next) next
```

**Comment by facundominguez** [comment]
> LBS.unpack (toLazyByteString acc)

I would use `show` instead of `unpack`, in case of encoding differences or characters that are harder to spot in a terminal.

**Comment by qaristote** [comment]
Shouldn't we also decorate error messages for other parts of the code that may fail ? Basically any call that does IO I guess. But then it sounds very tedious to add error-catching to every single call inside the library. 

Similarly, what about the `Z3` backend ?

**Comment by facundominguez** [comment]
> Shouldn't we also decorate error messages for other parts of the code that may fail ?

You could do so, but I'd only care if we have some way to trigger them.

> Similarly, what about the Z3 backend ?

If there is some way to make it fail, and the error is unclear, then it is worth to improve it. Otherwise, I wouldn't bother.

**Comment by facundominguez** [comment]
I contributed some simplifications. Feel free to merge if they look good to you.

**Review by facundominguez — APPROVED** [review]

**Pull request merged**